### PR TITLE
Track presence state per-device and combine to a user state.

### PR DESCRIPTION
**Pull request by clokep**
Tracks presence on an individual per-device basis and combines the per-device state into a per-user state. This should help in situations where a user has two devices with conflicting status (e.g. one is syncing with unavailable and one is syncing with online).

The tie-breaking is done by priority:

    BUSY > ONLINE > UNAVAILABLE > OFFLINE

Builds on #16170, #16171, #16172.

### CODE REVIEW

File: synapse/handlers/presence.py [hunk 1]
```diff
@@ -755,6 +772,11 @@ def run_persister() -> Awaitable[None]:
         self._event_pos = self.store.get_room_max_stream_ordering()
         self._event_processing = False
 
+        # The per-device presence state, maps user to devices to per-device presence state.
+        self.user_to_device_to_current_state: Dict[
+            str, Dict[Optional[str], UserDevicePresenceState]
+        ] = {}
```
Comment (clokep) [inline synapse/handlers/presence.py:778]: We don't currently remove things from this dictionary which means that once a user goes "online" it probably can't idle back to unavailable right now. (I need to add a test for this.) This is because we would consider the last activity from e.g. an unavailable sync loop, but the online state of a no-longer existing sync loop.

**Comment by clokep** [comment]
@erikjohnston Can you take a look at this and see if I'm wildly missing something / breaking the world?

File: synapse/handlers/presence.py [hunk 2]
```diff
@@ -926,6 +948,35 @@ async def _handle_timeouts(self) -> None:
...
+        new_states = []
+        for user_id in users_to_check:
+            user_devices = self.user_to_device_to_current_state.get(user_id, {})
+            # Only keep the devices that have been seen recently.
+            new_user_devices = {
+                device_id: device
+                for device_id, device in user_devices.items()
+                if now - max(device.last_user_sync_ts, device.last_active_ts)
+                <= IDLE_TIMER
+            }
```
[the remark this answers is not in the document]
Comment (clokep) [reply synapse/handlers/presence.py:961]: Isn't that what's happening? Old devices are considered idle and removed from the list, then `_combine_device_states` defaults to `OFFLINE` if there's no devices?

[the remark this answers is not in the document]
Comment (erikjohnston) [reply synapse/handlers/presence.py:961]: Ooooooooooh, sorry I didn't see the `last_user_sync_ts` there. I think you want to compare `device.last_user_sync_ts` with `SYNC_ONLINE_TIMEOUT` rather than `IDLE_TIMER` for that?

[the remark this answers is not in the document]
Comment (clokep) [reply synapse/handlers/presence.py:961]: Right, I think I combined the logic of two spots incorrectly. I think we want to be using the `IDLE_TIMER`, not `SYNC_ONLINE_TIMEOUT`? The data is still valid until we consider it to go back to idle; I think?

[the remark this answers is not in the document]
Comment (clokep) [reply synapse/handlers/presence.py:961]: Also we shouldn't be timing out `BUSY` devices, I think.

File: synapse/handlers/presence.py (hunk 1 above)
Comment (clokep) [reply synapse/handlers/presence.py:778]: I fixed this so that things are now timed out.

File: synapse/handlers/presence.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (clokep) [reply synapse/handlers/presence.py:962]: I've now cleaned this up to always use `IDLE_TIMER` and to handle `BUSY`.

File: changelog.d/16066.bugfix [hunk 3]
```diff
@@ -0,0 +1,2 @@
+Fix a long-standing bug where multi-device accounts could cause high load due to presence.
+
```
Comment (erikjohnston) [inline changelog.d/16066.bugfix:2]:
```suggestion
```

File: synapse/handlers/presence.py [hunk 4]
```diff
@@ -490,7 +502,10 @@ async def user_syncing(
             # what the spec wants: see comment in the BasePresenceHandler version
             # of this function.
             await self.set_state(
-                UserID.from_string(user_id), {"presence": presence_state}, True
+                UserID.from_string(user_id),
+                device_id,
+                {"presence": presence_state},
+                True,
```
Comment (erikjohnston) [inline synapse/handlers/presence.py:508]: Sorry, while we're here can we do these as keyword arguments please?

File: synapse/handlers/presence.py [hunk 5]
```diff
@@ -1249,6 +1308,27 @@ async def set_state(
 
         prev_state = await self.current_state_for_user(user_id)
 
+        # Always update the device specific information.
+        device_state = self.user_to_device_to_current_state.setdefault(
+            user_id, {}
+        ).setdefault(
+            device_id,
```
Comment (erikjohnston) [inline synapse/handlers/presence.py:1315]: Do we want to do this if `device_id` is None? As nothing will ever overwrite it?

File: synapse/handlers/presence.py [hunk 6]
```diff
@@ -2007,6 +2088,45 @@ def handle_update(
...
+
+    Return:
+        The combined presence state.
+    """
+
+    # Based on (all) the user's devices calculate the new presence state.
+    presence = PresenceState.OFFLINE
+    last_active_ts = -1
+
+    # Find the device to use presen priority device based on the presence priority, but tie-break with how recently the device has synced.
```
Comment (erikjohnston) [inline synapse/handlers/presence.py:2118]: You're using `last_active_ts` here rather than `lasy_sync_ts`?

(Shouldn't this have a line break?)

File: synapse/handlers/presence.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (realtyem) [reply synapse/handlers/presence.py:962]: I think I have to go with Erik on this one. The only state that should be discarded is offline, logically speaking. I can have my laptop idle for hours and intermittently connect with my mobile. What happens to my laptop status when my mobile disconnects if the timeout fires during that? Or even well before? What would I show up as then?

File: synapse/handlers/presence.py (hunk 5 above)
Comment (realtyem) [reply synapse/handlers/presence.py:1315]: I think in some unit tests, the device_id is set to be `None`. That shouldn't occur in the real/outside world though, would it?

File: synapse/handlers/presence.py [hunk 7]
```diff
@@ -1249,6 +1308,27 @@ async def set_state(
...
+            device_id,
+            UserDevicePresenceState(
+                user_id,
+                device_id,
+                presence,
+                last_active_ts=self.clock.time_msec(),
+            ),
+        )
+        device_state.state = presence
+        device_state.last_active_ts = self.clock.time_msec()
```
Comment (realtyem) [inline synapse/handlers/presence.py:1324]: I would say have this be either the previous `UserPresenceState.last_active_ts` here, and `0` right above as the default(or the -1 you have in other places). If this call to `set_state()` is being provided with `online` as the state to set, it will be updated regardless(currently, every time). But if the state to set is something else, it should be pulled from the previous state. Whatever you do with it, don't use the current time as that would violate how `currently_active` is supposed to be used(and would fire in `handle_update()`.

File: synapse/handlers/presence.py [hunk 8]
```diff
@@ -1962,6 +2042,7 @@ def handle_update(
     # If the users are ours then we want to set up a bunch of timers
     # to time things out.
     if is_mine:
+        # TODO Maybe don't do this if currently active?
```
Comment (realtyem) [inline synapse/handlers/presence.py:2045]: Fortunately, these aren't 'active timers', just a timestamp placeholder checked over every 5 seconds for new stuff. I thought it was contributing to the overhead of presence, it's not. At least not directly. I think it should be fine to keep as-is.

File: synapse/handlers/presence.py [hunk 9]
```diff
@@ -926,6 +948,36 @@ async def _handle_timeouts(self) -> None:
...
+
+                new_states.append(
+                    self.user_to_current_state.get(
+                        user_id, UserPresenceState.default(user_id)
+                    ).copy_and_replace(state=presence)
+                )
+
+                self.user_to_device_to_current_state[user_id] = new_user_devices
+        if new_states:
+            await self._update_states(new_states)
```
Comment (realtyem) [inline synapse/handlers/presence.py:979]: If you can, try and find a way to consolidate this into the `_update_states` call at the end of the function please. Every call to `_update_states` is a source of overhead(database and sometimes federation).

File: synapse/handlers/presence.py (hunk 4 above)
Comment (realtyem) [reply synapse/handlers/presence.py:508]: (Below for the similar place in the `PresenceHandler` version too please)

File: synapse/handlers/presence.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (clokep) [reply synapse/handlers/presence.py:962]: > won't this logic remove that device from the list and hence mark the user as offline?

Yes -- my understanding is that this is equivalent to the old logic. See:

https://github.com/matrix-org/synapse/blob/266bc1e49f60b6b591fa2189fbdfd73f9562d5db/synapse/handlers/presence.py#L1992-L2000

File: synapse/handlers/presence.py (hunk 5 above)
Comment (clokep) [reply synapse/handlers/presence.py:1315]: > As nothing will ever overwrite it?

Why would nothing ever overwrite it? There are few types of users which have access tokens without associated device IDs:

* Guest users
* Appservice users
* Access tokens created using admin APIs

File: synapse/handlers/presence.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (realtyem) [reply synapse/handlers/presence.py:962]: That is for users who are not syncing though? This other is for those still syncing but have gone idle, right?

[the remark this answers is not in the document]
Comment (clokep) [reply synapse/handlers/presence.py:962]: > That is for users who are not syncing though? This other is for those still syncing but have gone idle, right?

Aren't those the same thing? If you're not syncing then you have gone idle since the idle timeout is longer than the sync timeout?

[the remark this answers is not in the document]
Comment (erikjohnston) [reply synapse/handlers/presence.py:962]: Ah, no. If the `/sync` loop has `set_presence=unavailable` on it then it won't bump the active time but will bump the sync time.

[the remark this answers is not in the document]
Comment (realtyem) [reply synapse/handlers/presence.py:962]: Not at all. A sync can stop because the client has disconnected, which *should* mark them `offline`. A client that is still connected and syncing, but is idle, isn't the same thing. It just means they aren't interacting with anything. I can leave my client minimized so I can receive notifications, which *should* make me idle. (Although I feel like the spec could probably help right about now. This is all based on interpretation and point of view).

In other words, if they meant the same thing then why do they both exist?

File: synapse/handlers/presence.py (hunk 9 above)
Comment (clokep) [reply synapse/handlers/presence.py]: This code all got moved into `handle_timeout()`, so resolving.

File: synapse/handlers/presence.py [hunk 10]
```diff
@@ -1871,34 +2032,51 @@ def handle_timeout(
...
+            # If there are have been no sync for a while (and none ongoing),
+            # set presence to offline
+            if device_id not in syncing_device_ids:
+                # If the user has done something recently but hasn't synced,
+                # don't set them as offline.
+                sync_or_active = max(
+                    device_state.last_sync_ts, device_state.last_active_ts
+                )
+                if now - sync_or_active > SYNC_ONLINE_TIMEOUT:
+                    device_state.state = PresenceState.OFFLINE
```
Comment (clokep) [inline synapse/handlers/presence.py:2057]: Should we delete the device (which is equivalent to setting it offline, essentially)?

File: synapse/handlers/presence.py [hunk 11]
```diff
@@ -1871,34 +2052,63 @@ def handle_timeout(
...
+                # will be stuck in that state.
+                online_timeout = (
+                    BUSY_ONLINE_TIMEOUT
+                    if device_state.state == PresenceState.BUSY
+                    else SYNC_ONLINE_TIMEOUT
+                )
+                if now - sync_or_active > online_timeout:
+                    # Mark the device as going offline.
+                    offline_devices.append(device_id)
+                    device_changed = True
```
Comment (clokep) [inline synapse/handlers/presence.py:2089]: Without this a device could get stuck in `busy` and cause the entire account to be stuck in `busy`. This would be annoying. I chose a somewhat arbitrary 1 hour for `BUSY_ONLINE_TIMEOUT`, but maybe that's too long?

See also https://github.com/matrix-org/matrix-spec-proposals/pull/3026/files#r1287453423

File: synapse/replication/http/presence.py
Comment (clokep) [inline synapse/replication/http/presence.py:1]: I believe this file is somewhat reasonable for forwards/backwards compatibility but haven't checked it to a great degree.

File: synapse/replication/tcp/commands.py [hunk 12]
```diff
@@ -267,27 +267,38 @@ class UserSyncCommand(Command):
...
 
     def to_line(self) -> str:
         return " ".join(
             (
                 self.instance_id,
                 self.user_id,
+                str(self.device_id),
                 "start" if self.is_syncing else "end",
                 str(self.last_sync_ms),
             )
```
Comment (clokep) [inline synapse/replication/tcp/commands.py:304]: I think changing this will mean that the presence stream will somewhat break until all workers are upgraded or downgraded?

File: tests/server.py [hunk 13]
```diff
@@ -999,8 +999,6 @@ def setup_test_homeserver(
     hs.tls_server_context_factory = Mock()
 
     hs.setup()
-    if homeserver_to_use == TestHomeServer:
-        hs.setup_background_tasks()
```
Comment (clokep) [inline tests/server.py:1003]: `hs.setup()` calls this anyway.

File: synapse/handlers/presence.py [hunk 14]
```diff
@@ -1227,18 +1372,40 @@ async def set_state(
...
         user_id = target_user.to_string()
+        now = self.clock.time_msec()
 
         prev_state = await self.current_state_for_user(user_id)
 
+        # Syncs do not override a previous presence of busy.
+        #
+        # XXX: This is *not* part of the MSC and is likely unneeded with multi-device support.
+        if prev_state.state == PresenceState.BUSY and is_sync:
+            presence = PresenceState.BUSY
```
Comment (clokep) [inline synapse/handlers/presence.py:1383]: I *think* this is a hack for lack of multi-device support.

Note that it was added in #12213 and fixed in #14393. (#9644 added initial busy status stuff for reference.)

File: synapse/handlers/presence.py (hunk 5 above)
Comment (clokep) [reply synapse/handlers/presence.py:1388]: I'm not super thrilled to have to support `None` in all of this code though I also ended up using it for the special pseudo-device created on-start.

File: synapse/handlers/presence.py (hunk 14 above)
Comment (dbkr) [reply synapse/handlers/presence.py:1383]: Yes, I distinctly remember this being quite clunky and confusing without multi-device support and wanting to switch to a multi-device presence model, so yes, I think you're correct in that this is essentially a hack.

Comment (clokep) [reply synapse/handlers/presence.py:1383]: Thanks! I suspect some clients that implement busy might depend on it though? We'll need to double check whenever this is stabilized.

File: synapse/handlers/presence.py [hunk 15]
```diff
@@ -958,17 +1080,33 @@ async def bump_presence_active_time(self, user: UserID) -> None:
...
         bump_active_time_counter.inc()
 
-        prev_state = await self.current_state_for_user(user_id)
+        now = self.clock.time_msec()
 
-        new_fields: Dict[str, Any] = {"last_active_ts": self.clock.time_msec()}
-        if prev_state.state == PresenceState.UNAVAILABLE:
-            new_fields["state"] = PresenceState.ONLINE
+        # Update the device information & mark the device as online if it was
+        # unavailable.
```
Comment (erikjohnston) [inline synapse/handlers/presence.py:1086]: I feel like this is a "what" comment rather than a "why" comment? Why can't we go from offline -> online? Presumably because we need to have a sync stream to transition out of offline, which we clearly don't have?

File: synapse/handlers/presence.py [hunk 16]
```diff
@@ -1080,12 +1196,23 @@ async def update_external_syncs_row(
...
+
             updates = []
-            if is_syncing and user_id not in process_presence:
-                if prev_state.state == PresenceState.OFFLINE:
+            if is_syncing and (user_id, device_id) not in process_presence:
+                if device_state.state == PresenceState.OFFLINE:
+                    # Mark the device as online since it is not syncing, this
+                    # might mark the user as online, so send an update.
+                    device_state.state = PresenceState.ONLINE
+                    presence = _combine_device_states(devices.values())
```
Comment (erikjohnston) [inline synapse/handlers/presence.py:1211]: I feel like a lot of this logic is a bit scattered around? I'd expect the state machine conditions to be in one place where we update the device info?

File: synapse/handlers/presence.py [hunk 17]
```diff
@@ -1988,6 +2205,46 @@ def handle_update(
...
+    PresenceState.BUSY: 4,
+    PresenceState.ONLINE: 3,
+    PresenceState.UNAVAILABLE: 2,
+    PresenceState.OFFLINE: 1,
+}
+
+
+def _combine_device_states(
+    device_states: Iterable[UserDevicePresenceState],
+) -> str:
```
Comment (erikjohnston) [inline synapse/handlers/presence.py:2218]:
```suggestion
) -> PresenceState:
```

no?

**Review by erikjohnston — COMMENTED** [review]
Seems to mostly make sense, though it is a huge PR. Left some comments, but I think we might need to sit down with it? I'm assuming there is now chance of splitting this PR up?

File: synapse/handlers/presence.py (hunk 15 above)
Comment (clokep) [reply synapse/handlers/presence.py:1086]: That's fair, I think the code is just a bit complicated and was attempting to put a succinct description of what's going on.

File: synapse/handlers/presence.py (hunk 16 above)
Comment (clokep) [reply synapse/handlers/presence.py:1211]: Yes, I did not love how this is scattered around. Will give it a think! 👍

File: synapse/handlers/presence.py (hunk 11 above)
Comment (clokep) [reply synapse/handlers/presence.py]: Spun out into #16174.

File: synapse/handlers/presence.py (hunk 15 above)
Comment (clokep) [reply synapse/handlers/presence.py:1086]: I think this is pretty simplified now, but we still do call `_combine_presence_state` in a few spots.

File: synapse/handlers/presence.py (hunk 16 above)
Comment (clokep) [reply synapse/handlers/presence.py]: I think this is pretty simplified now, but we still do call `_combine_presence_state` in a few spots.

File: synapse/handlers/presence.py (hunk 7 above)
Comment (clokep) [reply synapse/handlers/presence.py]: I think this code doesn't exist or got moved to a different PR, sorry. 😢

File: synapse/handlers/presence.py (hunk 5 above)
Comment (clokep) [reply synapse/handlers/presence.py:1384]: Let's continue this on https://github.com/matrix-org/synapse/pull/16171/files#r1303245932.

File: synapse/handlers/presence.py [hunk 18]
```diff
@@ -2032,6 +2179,46 @@ def handle_update(
...
+
+    # Find the device to use the presence state of based on the presence priority,
+    # but tie-break with how recently the device has been seen.
+    for device_state in device_states:
+        if (PRESENCE_BY_PRIORITY[device_state.state], device_state.last_active_ts) > (
+            PRESENCE_BY_PRIORITY[presence],
+            last_active_ts,
+        ):
+            presence = device_state.state
+            last_active_ts = device_state.last_active_ts
```
Comment (clokep) [inline synapse/handlers/presence.py:2217]: This could probably all be replaced with `max(device_states)` if we define the correct [comparison method](https://docs.python.org/3/reference/datamodel.html#object.__lt__).

Comment (DMRobertson) [reply synapse/handlers/presence.py:2217]: Or even just  `max` with an explicit `key` function.

**Review by erikjohnston — APPROVED** [review]
